source/compiler/sc5.c

```diff
@@ -423,3 +450,173 @@ int pc_geterrorwarnings()
+    d[k]=k;
+  for (k=0;k<m;k++)
+    d[k*n]=k;
+  //Step 3 and 4
+  for (i=1;i<n;i++) {
+    for (j=1;j<m;j++) {
+      //Step 5
+      cost= (tolower(s[i-1])!=tolower(t[j-1]));
+      //Step 6
+      d[j*n+i]=minimum(d[(j-1)*n+i]+1,d[j*n+i-1]+1,d[(j-1)*n+i-1]+cost);
+    } /* for */
+  } /* for */
+  distance=d[n*m-1];
+  free(d);
+  return distance;
+}
+
+static int find_closestsymbol_table(const char *name,const symbol *root,int symboltype,symbol **closestsym)
+{
+  int dist,closestdist=INT_MAX;
+  char symname[2*sNAMEMAX+16];
+  symbol *sym=root->next;
+  int ident,critdist;
+
+  assert(closestsym!=NULL);
+  *closestsym=NULL;
+  assert(name!=NULL);
+  critdist=strlen(name)/2;  /* for short names, allow only a single edit */
+  if (critdist>MAX_EDIT_DIST)
+    critdist=MAX_EDIT_DIST;
+  while (sym!=NULL) {
+    funcdisplayname(symname,sym->name);
+    ident=sym->ident;
+    if (symboltype==iARRAY && ident==iREFARRAY)
+      ident=iARRAY;     /* reference arrays match arrays */
+    else if (symboltype==iVARIABLE && (sym->ident==iCONSTEXPR || sym->ident==iREFERENCE || sym->ident==iARRAY || sym->ident==iREFARRAY))
+      ident=iVARIABLE;  /* when requesting variables, constants are also ok */
+    if ((symboltype==ident || (symboltype==iVARIABLE && ident==iFUNCTN)) && ((ident!=iFUNCTN && ident!=iLABEL) || (sym->usage & uDEFINE)!=0)) {
+      dist=levenshtein_distance(name,symname);
+      if (dist<closestdist && dist<=critdist) {
+        *closestsym=sym;
+        closestdist=dist;
+      } /* if */
+    } /* if */
+    sym=sym->next;
+  } /* while */
+  return closestdist;
+}
+
+static symbol *find_closestsymbol(const char *name,int symboltype)
+{
+  symbol *symloc,*symglb;
+  int distloc,distglb;
+
+  if (sc_status==statFIRST)
+    return NULL;
+  assert(name!=NULL);
+  if (strlen(name)==0)
+    return NULL;
+  distloc=find_closestsymbol_table(name,&loctab,symboltype,&symloc);
+  if (distloc<=1)
+    distglb=INT_MAX; /* don't bother searching in the global table */
+  else
+    distglb=find_closestsymbol_table(name,&glbtab,symboltype,&symglb);
+  return (distglb<distloc) ? symglb : symloc;
+}
+
+static constvalue *find_closestautomaton(const char *name)
+{
+  constvalue *ptr=sc_automaton_tab.first;
+  constvalue *closestmatch=NULL;
+  int dist,closestdist=INT_MAX;
+
+  assert(name!=NULL);
+  while (ptr!=NULL) {
+    if (strlen(ptr->name)>0) {
+      dist=levenshtein_distance(name,ptr->name);
+      if (dist<closestdist && dist<=MAX_EDIT_DIST) {
+        closestmatch=ptr;
+        closestdist=dist;
+        if (dist<=1)
+          break;
+      } /* if */
+    } /* if */
+    ptr=ptr->next;
+  } /* while */
+  return closestmatch;
+}
+
+static constvalue *find_closeststate(const char *name,int fsa)
+{
+  constvalue *ptr=sc_state_tab.first;
+  constvalue *closestmatch=NULL;
+  int dist,closestdist=INT_MAX;
+
+  assert(name!=NULL);
+  assert(closestmatch!=NULL);
+  while (ptr!=NULL) {
+    if (ptr->index==fsa && strlen(ptr->name)>0) {
```

Comment:
It would be nice if `find_closestautomaton` and this could be merged - is there a constant value of `->index` in `sc_automaton_tab`?  If so, `fsa` could be set to that, and an extra function deleted.